blockchain/indexers/flatutreexoproofindex.go

```diff
@@ -167,21 +174,40 @@ func (idx *FlatUtreexoProofIndex) ConnectBlock(dbTx database.Tx, block *btcutil.
 		return err
 	}
 
-	idx.mtx.Lock()
-	undoBlock, err := idx.utreexoState.state.Modify(adds, ud.AccProof.Targets)
-	idx.mtx.Unlock()
+	delHashes := make([]utreexo.Hash, 0, len(ud.LeafDatas))
+	for _, ld := range ud.LeafDatas {
+		if ld.IsUnconfirmed() {
+			continue
+		}
+		delHashes = append(delHashes, ld.LeafHash())
+	}
+
+	err = idx.storeUndoBlock(block.Height(),
+		uint64(len(adds)), ud.AccProof.Targets, delHashes)
 	if err != nil {
 		return err
 	}
 
-	idx.pStats.UpdateTotalDelCount(uint64(len(dels)))
-	idx.pStats.UpdateUDStats(false, ud)
+	err = idx.storeRoots(block.Height(), idx.utreexoState.state)
```

Comment:
Ok so since we have `UpdateData` available now, we don't necessarily need the roots to do the undo. However, since the pollard struct can't generate/use the updatedata, we can't get rid of the roots just yet.

So I think just keeping this as is for now and coming back to it later might be fine because everything consensus related is working. It's just a bit more straightforward to someone that'd be using this codebase... which is only me for now.